contrib/ivorysql_ora/src/builtin_functions/character_datatype_functions.c

```diff
@@ -1690,4 +1691,84 @@ text_position_prev(int start_pos, TextPositionState *state)
 	}
 
 	return pos;
+}
+
+/*
+ *	appendUTF16Escape
+ *
+ *	Converts a UTF-16 code unit to a Unicode escape sequence (\xxxx) 
+ * 	and appends it to the output string.
+ */
+static void 
+appendUTF16Escape(StringInfoData* outputString, uint16_t codePoint) {
+    char buffer[10];
+    snprintf(buffer, sizeof(buffer), "\\%04X", codePoint);
+    appendStringInfoString(outputString, buffer);
+}
+
+/********************************************************************
+ * ora_asciistr
+ *
+ * Purpose:
+ *	 It takes string as a argument, or an expression that resolves to 
+ * 	 a string, in any character set and returns an ASCII 
+ *   version of the string in the database character set. 
+ *   Non-ASCII characters are converted to the form \xxxx, 
+ *   where xxxx represents a UTF-16 code unit. 
+ ********************************************************************/
+Datum 
+ora_asciistr(PG_FUNCTION_ARGS) {
+	StringInfoData 	output;
+	text 			*str_arg = NULL;
+	char 			*str = NULL;
+    char			*end = NULL;
+	
+	initStringInfo(&output);
+	str_arg = PG_GETARG_TEXT_PP(0);
+	str = VARDATA_ANY(str_arg);
+	end = str + VARSIZE_ANY_EXHDR(str_arg);
+
+    while (str < end) {
+        unsigned char c = *str;
+        uint32_t codePoint;
+		
+		if (c == '\\') {
+			/* Handle backslash character */
+			appendUTF16Escape(&output, 0x005C);  // UTF-16 representation of backslash
+			str++;
+        } 
+        else if (c < 0x80) {
+            /* ASCII character */
+            appendStringInfoChar(&output, c);
+            str++;
+        } else if ((c & 0xE0) == 0xC0) {
+            /* Two-byte UTF-8 sequence */
+            codePoint = ((c & 0x1F) << 6) | (str[1] & 0x3F);
+            appendUTF16Escape(&output, (uint16_t) codePoint);
+            str += 2;
+        } else if ((c & 0xF0) == 0xE0) {
+            /* Three-byte UTF-8 sequence */
+            codePoint = ((c & 0x0F) << 12) | ((str[1] & 0x3F) << 6) | (str[2] & 0x3F);
+            appendUTF16Escape(&output, (uint16_t) codePoint);
```

Comment:
since a three-byte unicode sequence is being added, uint32_t should be used here. 
As a result, the second parameter of appendUTF16Escape() should be uint32_t too.

Reply:
* The three-byte UTF-8 sequence (0xE0) covers code points from U+0800 to U+FFFF which falls under the BMP and can be represented by a single uint16_t.

https://exploringjs.com/js/book/ch_unicode.html#utf-16
https://learn.microsoft.com/en-us/globalization/encoding/unicode-standard#surrogate-pairs